36_workflow-explore-extend-pull-request.Rmd

```diff
@@ -0,0 +1,140 @@
+
+This set of instructions suggests that you clone the fork, checkout the branch from which the PR was made, make any commits you wish, and then push. Any new commits you make will appear in the PR. And then you could merge.
+
+My main takeaway: maintainer can push to the branch of a fork associated with a PR.
+
+## My under-development workflow
+
+*work in progress*
+
+This combines ideas from the two above approaches, but with some tweaks. I have the start of a workflow from R, but it's incomplete, so will sketch with command line Git.
+
+Determine the name of the local PR branch. If PR came from non-`master` branch, use that branch name. Otherwise, you'll have to make something up, maybe based on the PR number and/or PR username.
+
+Determine the parent commit of the PR, i.e. the SHA in master the PR is based on.
+
+Create and check out a new local PR branch with the chosen name and chosen SHA as parent.
+
+```
+git checkout -b completions a1c46a8
+git checkout -b PR_BRANCH PARENT_SHA
+```
+
+Pull from the fork and branch associated with the PR into the local PR branch. Make a deliberate choice re: HTTPS or SSH, based on what you usually use. I show SSH here.
+
+```
+git pull --ff git@github.com:jimhester/readxl.git completions
+git pull --ff git@github.com:OWNER/REPO.git FORK_BRANCH
+```
+
+Experiment with the PR and make more commits.
+
+Push back to the fork branch associated with the PR.
+
+```
+git push git@github.com:jimhester/readxl.git HEAD:completions
+git push git@github.com:OWNER/REPO.git HEAD:FORK_BRANCH
+```
+
+Merge or squash-and-merge the updated PR from GitHub in the browser.
+
+Partial draft of R code to do the above
+
+```{r, eval = FALSE}
+library(gh)
+library(git2r)
+
+## the number of the pull request you want to work on
+pr <- 320
+## assuming wd = active project/package, gets OWNER/REPO
+repo_info <- gh_tree_remote(".")
+
+x <- gh(
+  "/repos/:owner/:repo/pulls/:number",
+  owner = repo_info$username,
+  repo = repo_info$repo,
+  number = pr
+)
+
+## get name of the fork's branch used for the PR
+(fork_branch <- x$head$ref)
+## determine name of local PR branch
+(pr_branch <- if (fork_branch == "master") paste0("pr-", pr) else fork_branch)
+
+## get the parent SHA for the PR
+
+## why is this not correct?!?!?
+# (sha <- x$base$sha)
+
+## alternative approach
+## GET all commits for the PR and parent SHA of the first commit
+y <- gh(
+  "/repos/:owner/:repo/pulls/:number/commits",
+  owner = repo_info$username,
+  repo = repo_info$repo,
+  number = pr
+)
+(sha <- purrr::pluck(y, list(1, "parents", 1, "sha")))
+
+## create and checkout the local PR branch, with correct parent
+b <- branch_create(commit = lookup(sha = sha), name = pr_branch)
+checkout(b)
+
+## haven't figured out how to pull / push to arbitrary remote with git2r
+## perhaps you'd have to actually add the fork as a remote?
+## doing with command line Git for now
```

Comment:
It doesn't look like git2r has an API that lets you use anonymous remotes, I think you need to explicitly add the remote then fetch, then do the branch, e.g. `remote_add()`, `fetch()`, `branch()` etc.